linode/linode-cli
# ref: Move build configuration from `setup.py` to `pyproject.toml`; apply various repo/CI refactors

## 📝 Description

This change makes the following build system improvements to the Linode CLI:

* Moving away from `setup.py` in favor of setuptools in `pyproject.toml`
* Moving away from the `baked_version` construct in favor of injecting the version into a new `version.py` file.
* Removing `requirements.txt` and `requirements-dev.txt` in favor of `pyproject.toml`
* Removing the logic to automatically bake and install bash completions to `/etc/bash_completion.d`
    * This was discussed internally as a team
* Removing the Jenkinsfile (no longer used)
* Updating GHA workflows and Dockerfiles accordingly
* Updating the OCI publish workflow to not rely on the `version` script (which was previously removed)

## ✔️ How to Test

The following test steps assume you have pulled down this PR locally.

### Testing Workflows

For convenience I tested all workflows on the `main` branch of my fork; they can be viewed here:
* [OCI Build & Publish](https://github.com/lgarber-akamai/linode-cli/actions/runs/8804593829/job/24165298002)
* [Release](https://github.com/lgarber-akamai/linode-cli/actions/runs/8804593828/job/24165298009)
* [Integration Tests (failing due to unrelated configuration issue)](https://github.com/lgarber-akamai/linode-cli/actions/runs/8804591132/job/24165287589) 
* [PR Integration Tests (failing due to unrelated issues)](https://github.com/lgarber-akamai/linode-cli/actions/runs/8804876351)

### Testing the Build & Install Process

1. Run the following command to build & install the CLI:

```
make install
```

2. Ensure the CLI has installed successfully and the installed version matches `v0.0.0.dev`:

```
linode-cli --version

# linode-cli 0.0.0.dev0
# Built from spec version 4.174.0
```

3. Install the CLI with an overridden version:

```
LINODE_CLI_VERSION=1.2.3 make install
```

4. Ensure the CLI has installed correctly and the installed version matches `v0.0.0.dev`:

```
linode-cli --version

# linode-cli 1.2.3
# Built from spec version 4.174.0
```

### Testing the Dockerfile Changes

1. Create a read-only GitHub Personal Access Token and export it under the `GITHUB_TOKEN` environment variable.

2. Run the following command to build the CLI OCI image:

```
docker build -t local-linode-cli --build-arg linode_cli_version=1.2.3 --build-arg github_token=$GITHUB_TOKEN .
```

3. Attempt to execute a CLI command, for example:

```
docker run --rm -it -e LINODE_CLI_TOKEN=$LINODE_TOKEN local-linode-cli linodes ls
```

4. Ensure the command runs as expected.

### Unit Testing

```
make testunit
```

### Integration Testing

```
make testint
```
